Create Kubernetes 1.24 Release Blog Article (#32867)
kubernetes/website: content/en/blog/_posts/2022-4-26-kubernetes-release-1.24.md

## Major Themes

### Dockershim Removed from kubelet
After an initial deprecation in v1.20, the dockershim has been removed from the kubelet in favor of runtimes that comply with the Container Runtime Interface (CRI) designed for Kubernetes. From v1.24 and up, if you are currently relying on Docker Engine as your container runtime, you will need to either use one of the other supported runtimes (such as containerd or CRI-O) or use cri-dockerd. For more information about ensuring your cluster is ready for this removal, please see [this guide](/blog/2022/03/31/ready-for-dockershim-removal/). 
  
### CNI Version-Related Breaking Change
Before you upgrade to Kubernetes 1.24, if you use a [Container Network Interface (CNI)](https://github.com/containernetworking/cni) network plugin that complies with CNI versions earlier than v1.0.0 and you plan to use any of the following container runtimes, you must specify the CNI version as described in [Declaring the CNI version of your plugin](TBD). Otherwise, your network will break. 

This issue applies to the following container runtimes:
* containerd v1.6.0 and later

> * CRI-O 1.24 and later

Reviewer: This isn't a Major Theme for the release. It **is** a key detail; I'd mention it after Major Themes and before  Other Updates, with its own 2nd-level heading.

### Beta APIs Off by Default 
[New beta APIs will not be enabled in clusters by default](https://github.com/kubernetes/enhancements/issues/3136). Existing beta APIs and new versions of existing beta APIs, will continue to be enabled by default.
  
### OpenAPI v3
Kubernetes 1.24 offers beta support for publishing its APIs as [OpenAPI v3](https://github.com/kubernetes/enhancements/issues/2896). 

### Signing Release Artifacts

> [Release artifacts are signed](https://github.com/kubernetes/enhancements/issues/3031). 

Reviewer: Could we have:
```suggestion
Release artifacts are [signed](https://github.com/kubernetes/enhancements/issues/3031)
and there is experimental support for [verifying image signatures](/docs/tasks/administer-cluster/verify-signed-images/).
```
?

See https://deploy-preview-31225--kubernetes-io-main-staging.netlify.app/docs/tasks/administer-cluster/verify-signed-images/ for a preview of that linked task page.

Are _all_ artifacts signed, or just some?

  
### Storage Capacity and Volume Expansion Are Generally Available
[Storage capacity tracking](https://github.com/kubernetes/enhancements/issues/1472) supports exposing currently available storage capacity via CSIStorageCapacity objects and enhances scheduling of pods that use CSI volumes with late binding.

[Volume expansion](https://github.com/kubernetes/enhancements/issues/284) add support for resizing existing persistent volumes. 

### NonPreemptingPriority to Stable
This feature adds [a new option to PriorityClasses](https://github.com/kubernetes/enhancements/issues/902), which can enable or disable pod preemption. 

> ### In-Tree CSI Providers

Reviewer: ```suggestion
### Storage plugin migration
```
There is no such thing as an “in tree CSI provider”.

> [Migrate the internals of the in-tree plugins](https://github.com/kubernetes/enhancements/issues/625) to call out to CSI Plugins while maintaining the original API. [Azure Disk](https://github.com/kubernetes/enhancements/issues/1490) and [OpenStack Cinder](https://github.com/kubernetes/enhancements/issues/1489) have both been migrated.

Reviewer: ```suggestion
There is work under way to [migrate the internals of in-tree storage plugins](https://github.com/kubernetes/enhancements/issues/625) to call out to CSI Plugins,
while maintaining the original API.
The [Azure Disk](https://github.com/kubernetes/enhancements/issues/1490)
and [OpenStack Cinder](https://github.com/kubernetes/enhancements/issues/1489) plugins
have both been migrated.
```

> ### Graduations to Stable

Reviewer: Should this be under the Other Updates heading?

This release saw fifteen enhancements promoted to stable: 

* [Container Storage Interface (CSI) Volume Expansion](https://github.com/kubernetes/enhancements/issues/284)

> * [PodOverhead](https://github.com/kubernetes/enhancements/issues/688): Account for resources tied to the pod sandbox, but not specific containers. 

Reviewer: ```suggestion
* [Pod Overhead](https://github.com/kubernetes/enhancements/issues/688): Account for resources tied to the pod sandbox, but not specific containers.
```
This is a phrase, not an API.

* [Add non-preempting option to PriorityClasses](https://github.com/kubernetes/enhancements/issues/902)
* [Storage Capacity Tracking](https://github.com/kubernetes/enhancements/issues/1472) 
* [OpenStack Cinder In-Tree to CSI Driver Migration](https://github.com/kubernetes/enhancements/issues/1489)
* [Azure Disk In-Tree to CSI Driver Migration](https://github.com/kubernetes/enhancements/issues/1490)
* [Efficient Watch Resumption](https://github.com/kubernetes/enhancements/issues/1904): Watch can be efficiently resumed after kube-apiserver reboot. 

> * [Service Type=LoadBalancer Class](https://github.com/kubernetes/enhancements/issues/1959): Introduce a new Service annotation `service.kubernetes.io/load-balancer-class` that allows multiple implementations of Service Type=LoadBalancer in a cluster.

Reviewer: How about:
```suggestion
* [Service Type=LoadBalancer Class Field](https://github.com/kubernetes/enhancements/issues/1959): Introduce a new Service annotation `service.kubernetes.io/load-balancer-class` that allows multiple implementations of `type: LoadBalancer` Services in the same cluster.
```

> * [Indexed Job Semantics in Job API](https://github.com/kubernetes/enhancements/issues/2214): Add a completion index to Pods of Jobs with fixed completion count. 

Reviewer: ```suggestion
* [Indexed Job](https://github.com/kubernetes/enhancements/issues/2214): Add a completion index to Pods of Jobs with fixed completion count.
```

(I'm using the title of the KEP instead of the title of the tracking issue)

> * [batch/v1: Add Suspend Field to Jobs API](https://github.com/kubernetes/enhancements/issues/2232): Add a suspend field to the Jobs API to allow orchestrators to create jobs with more control over when pods are created.

Reviewer: ```suggestion
* [Add Suspend Field to Jobs API](https://github.com/kubernetes/enhancements/issues/2232): Add a suspend field to the Jobs API to allow orchestrators to create jobs with more control over when pods are created.
```

> * [Pod Affinity NamespaceSelector](https://github.com/kubernetes/enhancements/issues/2249): Adds namespaceSelector to pod affinity/anti-affinity spec. 

Reviewer: ```suggestion
* [Pod Affinity NamespaceSelector](https://github.com/kubernetes/enhancements/issues/2249): Add a `namespaceSelector` field for to pod affinity/anti-affinity spec.
```

* [Leader Migration for Controller Managers](https://github.com/kubernetes/enhancements/issues/2436): kube-controller-manager and cloud-controller-manager can apply new controller-to-controller-manager assignment in HA control plane without downtime.

> * [CSR Duration](https://github.com/kubernetes/enhancements/issues/2784): Extend the Certificates API with a mechanism to allow clients to request a specific duration for the issued certificate.

Reviewer: ```suggestion
* [CSR Duration](https://github.com/kubernetes/enhancements/issues/2784): Extend the CertificateSigningRequest API with a mechanism to allow clients to request a specific duration for the issued certificate.
```

> * [Beta APIs are off by Default](https://github.com/kubernetes/enhancements/issues/3136)

Reviewer: This is not an actual graduation to stable. I'd remove it.

---
layout: blog
title: "Kubernetes 1.24: Stargazer"
date: 2022-05-03
slug: kubernetes-1-24-release-announcement
---

# Kubernetes 1.24

**Authors**: [Kubernetes 1.24 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.24/release-team.md)

We are excited to announce the release of Kubernetes 1.24, the first release of 2022!

This release consists of 46 enhancements: fifteen enhancements have graduated to stable, fifteen enhancements are moving to beta, and thirteen enhancements are entering alpha. Also, two features have been deprecated, and the [dockershim has been removed](https://kubernetes.io/blog/2022/02/17/dockershim-faq/).     

## Major Themes

### Dockershim Removed from kubelet

> After an initial deprecation in v1.20, the dockershim has been removed from the kubelet in favor of runtimes that comply with the Container Runtime Interface (CRI) designed for Kubernetes. From v1.24 and up, if you are currently relying on Docker Engine as your container runtime, you will need to either use one of the other supported runtimes (such as containerd or CRI-O) or use cri-dockerd. For more information about ensuring your cluster is ready for this removal, please see [this guide](/blog/2022/03/31/ready-for-dockershim-removal/). 

Reviewer 2: We're mixing referring to Kubernetes versions with a leading "v" and not. [Generally we don't use the leading "v".](https://kubernetes.io/blog/2021/12/07/kubernetes-1-23-release-announcement/)
```suggestion
After an initial deprecation in Kubernetes 1.20, the dockershim has been removed from the kubelet in favor of runtimes that comply with the Container Runtime Interface (CRI) designed for Kubernetes. From Kubernetes 1.24 and up, if you are currently relying on Docker Engine as your container runtime, you will need to either use one of the other supported runtimes (such as containerd or CRI-O) or use cri-dockerd. For more information about ensuring your cluster is ready for this removal, please see [this guide](/blog/2022/03/31/ready-for-dockershim-removal/). 
```

Reviewer: We prefer to link to the title rather than writing “click here”, “this guide” etc.

How about this wording (and perhaps also update the release notes if appropriate):
```suggestion
After an initial deprecation in Kubernetes 1.20, the dockershim has been removed
from the kubelet in favor of runtimes that comply with the Container Runtime Interface (CRI)
designed for Kubernetes.
From Kubernetes 1.24 and up, if you were previously relying on Docker Engine as your container
runtime, you will need to either switch to one a compatible runtime (such as containerd
or CRI-O), or use a third party adapter such as `cri-dockerd`. For more information,
see [Is Your Cluster Ready for v1.24?](/blog/2022/03/31/ready-for-dockershim-removal/).
```

  
### CNI Version-Related Breaking Change
Before you upgrade to Kubernetes 1.24, if you use a [Container Network Interface (CNI)](https://github.com/containernetworking/cni) network plugin that complies with CNI versions earlier than v1.0.0 and you plan to use any of the following container runtimes, you must specify the CNI version as described in [Declaring the CNI version of your plugin](TBD). Otherwise, your network will break. 

This issue applies to the following container runtimes:
* containerd v1.6.0 and later
* CRI-O 1.24 and later
* cri-dockerd [vTBD] and later

### Beta APIs Off by Default 
[New beta APIs will not be enabled in clusters by default](https://github.com/kubernetes/enhancements/issues/3136). Existing beta APIs and new versions of existing beta APIs, will continue to be enabled by default.
  
### OpenAPI v3
Kubernetes 1.24 offers beta support for publishing its APIs as [OpenAPI v3](https://github.com/kubernetes/enhancements/issues/2896). 

### Signing Release Artifacts

> [Release artifacts are signed](https://github.com/kubernetes/enhancements/issues/3031). 

Reviewer 2: Maybe some verbage here about why this is important?

Reviewer 3: Release artifacts are now signed in order to prevent possible man-in-the-middle attacks?

Reviewer 4: Maybe a reference to the [KEP 
3027](https://github.com/kubernetes/enhancements/issues/3027) would be good or stating increasing Supply Chain Security and Verifiable Release Artifacts?

* [Beta APIs are off by Default](https://github.com/kubernetes/enhancements/issues/3136)
* [Dockershim Removal](https://github.com/kubernetes/enhancements/issues/2221)

### gRPC Probes Graduate to Beta
With Kubernetes 1.24, the [gRPC probes functionality](https://github.com/kubernetes/enhancements/issues/2727) has entered beta and is available by default. You can now [configure startup, liveness, and readiness probes](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes) for your gRPC app natively within Kubernetes, without exposing an HTTP endpoint or
using an extra executable.

### Kubelet Credential Provider graduates to Beta
Released as Alpha in Kubernetes 1.20, the [Kubelet Credential Provider](https://github.com/kubernetes/enhancements/issues/2133) has now graduated to Beta. This allows the kubelet to dynamically retrieve credentials for a container image registry using exec plugins, communicating through stdio using Kubernetes versioned APIs, rather than storing them statically on disk.

### Contextual Logging in Alpha

> Kubernetes v1.24 has introduced [contextual logging](https://github.com/kubernetes/enhancements/issues/3077) that enables the caller of a function to control all aspects of logging (output formatting, verbosity, additional values and names).

Reviewer 2: ```suggestion
Kubernetes 1.24 has introduced [contextual logging](https://github.com/kubernetes/enhancements/issues/3077) that enables the caller of a function to control all aspects of logging (output formatting, verbosity, additional values and names).
```

### Avoiding Collisions in IP allocation to Services
Kubernetes 1.24 introduces a new Feature Gate `ServiceIPStaticSubrange` that allows users to [soft-reserve a range for static IP address assignments](https://github.com/kubernetes/enhancements/issues/3070). If you use the new feature (which you must manually enable), your cluster prefers automatic assignment from the remaining pool of Service IP addresses, reducing the risk of a collision.
  

>   

Reviewer 2: Why the double space?

Reviewer 3: I was told to always put double spaces after # headings in markdown

Reviewer: You can put 2 blank lines here. The preceding item is actually a paragraph not a heading, though.

As a convention, we usually leave a blank line after headings that use `#`, `##` etc as prefix.

## Other Updates
### Graduated to Stable
* To do: Add all enhancements that graduated to stable

### Major Changes
* To do: Add information about Dockershim

### Release Notes
Check out the full details of the Kubernetes 1.24 release in our [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md).
    
### Availability

Kubernetes 1.24 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.24.0).
To get started with Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run local 
Kubernetes clusters using containers as “nodes”, with [kind](https://kind.sigs.k8s.io/).
You can also easily install 1.24 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/).

### Release Team

This release would not have been possible without the combined efforts of a committed individuals comprising the Kubernetes 1.24 release team. This team came together to deliver all of the components that go into each Kubernetes release, including code, documentation, release notes, and more. 

> Special thanks to James Laverack, our release lead, for guiding us through a successful release cycle, and to all of the members of the release team for the time and effort they put in to deliver the v1.24 release for the Kubernetes community. 

Reviewer 2: You're gonna make me blush. 😊 

But also
```suggestion
Special thanks to James Laverack, our release lead, for guiding us through a successful release cycle, and to all of the members of the release team for the time and effort they put in to deliver the 1.24 release for the Kubernetes community. 
```

* [Efficient Watch Resumption](https://github.com/kubernetes/enhancements/issues/1904): Watch can be efficiently resumed after kube-apiserver reboot. 
* [Service Type=LoadBalancer Class Field](https://github.com/kubernetes/enhancements/issues/1959): Introduce a new Service annotation `service.kubernetes.io/load-balancer-class` that allows multiple implementations of `type: LoadBalancer` Services in the same cluster.
* [Indexed Job](https://github.com/kubernetes/enhancements/issues/2214): Add a completion index to Pods of Jobs with fixed completion count.
* [Add Suspend Field to Jobs API](https://github.com/kubernetes/enhancements/issues/2232): Add a suspend field to the Jobs API to allow orchestrators to create jobs with more control over when pods are created.
* [Pod Affinity NamespaceSelector](https://github.com/kubernetes/enhancements/issues/2249): Add a `namespaceSelector` field for to pod affinity/anti-affinity spec.
* [Leader Migration for Controller Managers](https://github.com/kubernetes/enhancements/issues/2436): kube-controller-manager and cloud-controller-manager can apply new controller-to-controller-manager assignment in HA control plane without downtime.

> * [CSR Duration](https://github.com/kubernetes/enhancements/issues/2784): Extend the CertificateSigningRequest API with a mechanism to allow clients to request a specific duration for the issued certificate.

Reviewer: I recommend that we write all of these in the same case convention. I recommend sentence case, but Title Case is OK too.

Author: Fixed it - I'm partial to title case so I went with that.

* [Beta APIs are off by Default](https://github.com/kubernetes/enhancements/issues/3136)
* [Dockershim Removal](https://github.com/kubernetes/enhancements/issues/2221)

### gRPC Probes Graduate to Beta
With Kubernetes 1.24, the [gRPC probes functionality](https://github.com/kubernetes/enhancements/issues/2727) has entered beta and is available by default. You can now [configure startup, liveness, and readiness probes](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes) for your gRPC app natively within Kubernetes, without exposing an HTTP endpoint or
using an extra executable.

### Kubelet Credential Provider graduates to Beta

> Released as Alpha in Kubernetes 1.20, the [Kubelet Credential Provider](https://github.com/kubernetes/enhancements/issues/2133) has now graduated to Beta. This allows the kubelet to dynamically retrieve credentials for a container image registry using exec plugins, communicating through stdio using Kubernetes versioned APIs, rather than storing them statically on disk.

Reviewer: ```suggestion
Originally released as Alpha in Kubernetes 1.20, the kubelet's support for
[image credential providers](/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/)
has now graduated to Beta.
This allows the kubelet to dynamically retrieve credentials for a container image registry
using exec plugins, rather than storing credentials on the node's filesystem.
```
(the task page has more detail so I think it's OK to omit the extra implementation information from this article)

Reviewer: It's OK to link to the tracking issue as well, but we should definitely make the link to the docs more prominent. Few readers care about the tracking issue.

### Contextual Logging in Alpha
Kubernetes 1.24 has introduced [contextual logging](https://github.com/kubernetes/enhancements/issues/3077) that enables the caller of a function to control all aspects of logging (output formatting, verbosity, additional values and names).

### Avoiding Collisions in IP allocation to Services

> Kubernetes 1.24 introduces a new Feature Gate `ServiceIPStaticSubrange` that allows users to [soft-reserve a range for static IP address assignments](https://github.com/kubernetes/enhancements/issues/3070). If you use the new feature (which you must manually enable), your cluster prefers automatic assignment from the remaining pool of Service IP addresses, reducing the risk of a collision.

Reviewer: I recommend linking to the docs rather than to the enhancement tracking issue:
```suggestion
Kubernetes 1.24 introduced a new opt-in feature that allows you to
[soft-reserve a range for static IP address assignments](/docs/concepts/services-networking/service/#service-ip-static-sub-range)
to Services.
If you use the new feature (which you must manually enable), your cluster prefers automatic
assignment from the remaining pool of Service IP addresses, reducing the risk of a collision.
```

This release consists of 46 enhancements: fifteen enhancements have graduated to stable, fifteen enhancements are moving to beta, and thirteen enhancements are entering alpha. Also, two features have been deprecated, and the [dockershim has been removed](/dockershim).     

## Major Themes

### Dockershim Removed from kubelet

After an initial deprecation in Kubernetes 1.20, the dockershim has been removed from the kubelet in favor of runtimes that comply with the Container Runtime Interface (CRI) designed for Kubernetes. From Kubernetes 1.24 and up, if you are currently relying on Docker Engine as your container runtime, you will need to either use one of the other supported runtimes (such as containerd or CRI-O) or use cri-dockerd. For more information about ensuring your cluster is ready for this removal, please see [this guide](/blog/2022/03/31/ready-for-dockershim-removal/). 

### Beta APIs Off by Default 

[New beta APIs will not be enabled in clusters by default](https://github.com/kubernetes/enhancements/issues/3136). Existing beta APIs and new versions of existing beta APIs, will continue to be enabled by default.
  
### OpenAPI v3

> Kubernetes 1.24 offers beta support for publishing its APIs as [OpenAPI v3](https://github.com/kubernetes/enhancements/issues/2896). 

Reviewer 5: Slight rewording to clarify that we're using the format/spec.
```suggestion
Kubernetes 1.24 offers beta support for publishing its APIs in the [OpenAPI v3 format](https://github.com/kubernetes/enhancements/issues/2896). 
```

### Signing Release Artifacts

> Release artifacts are [signed](https://github.com/kubernetes/enhancements/issues/3031)

Reviewer 5: I think it'd be worthwhile calling out the signing utility we are using for transparency.
```suggestion
Release artifacts are [signed](https://github.com/kubernetes/enhancements/issues/3031) using [cosign](https://github.com/sigstore/cosign) signatures
```

and there is experimental support for [verifying image signatures](/docs/tasks/administer-cluster/verify-signed-images/).
  
### Storage Capacity and Volume Expansion Are Generally Available

> [Storage capacity tracking](https://github.com/kubernetes/enhancements/issues/1472) supports exposing currently available storage capacity via CSIStorageCapacity objects and enhances scheduling of pods that use CSI volumes with late binding.

Reviewer 5: ```suggestion
[Storage capacity tracking](https://github.com/kubernetes/enhancements/issues/1472) supports exposing currently available storage capacity via [CSIStorageCapacity objects](https://kubernetes.io/docs/concepts/storage/storage-capacity/#api) and enhances scheduling of pods that use CSI volumes with late binding.
```

> [Volume expansion](https://github.com/kubernetes/enhancements/issues/284) add support for resizing existing persistent volumes. 

Reviewer 5: ```suggestion
[Volume expansion](https://github.com/kubernetes/enhancements/issues/284) adds support for resizing existing persistent volumes. 
```

using an extra executable.

### Kubelet Credential Provider graduates to Beta

Originally released as Alpha in Kubernetes 1.20, the kubelet's support for
[image credential providers](/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/)
has now graduated to Beta.
This allows the kubelet to dynamically retrieve credentials for a container image registry
using exec plugins, rather than storing credentials on the node's filesystem.

### Contextual Logging in Alpha

Kubernetes 1.24 has introduced [contextual logging](https://github.com/kubernetes/enhancements/issues/3077) that enables the caller of a function to control all aspects of logging (output formatting, verbosity, additional values and names).

### Avoiding Collisions in IP allocation to Services

Kubernetes 1.24 introduced a new opt-in feature that allows you to
[soft-reserve a range for static IP address assignments](/docs/concepts/services-networking/service/#service-ip-static-sub-range)
to Services.

> If you use the new feature (which you must manually enable), your cluster prefers automatic

Reviewer 5: Do we mean to say that with the manual enablement of this feature the cluster will opt for automatic assignment from a pool of Service IP addresses thereby reducing a risk of collision?

Author: Yes, I will rephrase.

assignment from the remaining pool of Service IP addresses, reducing the risk of a collision.
  
A Service `ClusterIP` can be assigned:

* dynamically, which means the cluster will automatically pick a free IP within the configured Service IP range.
* statically, which means the user will set one IP within the configured Service IP range.

Service `ClusterIP` are unique, hence, trying to create a Service with a `ClusterIP` that has already been allocated will return an error.

## CNI Version-Related Breaking Change

Before you upgrade to Kubernetes 1.24, please verify that you are using/upgrading to a container runtime that has been tested to work correctly with this release.

> The following container runtimes have been tested with Kubernetes 1.24:

Reviewer 5: I think this will need amendment because the versions of these container runtimes have not been released yet, to the best of my knowledge. I checked the containerd/CRI-O releases and containerd has just released 1.6.3 earlier this week & CRI-O had a minor version bump for 1.23 sometime back. We might want to consider rephrasing this after discussing it on the dockershim call today.

Reviewer 5: @mickeyboxell Please refer to the specific suggestion on the [task doc](https://github.com/kubernetes/website/pull/33287#discussion_r861966181) while amending.

Thank you @chris-short @nate-double-u @dims for your help on this.

---
layout: blog
title: "Kubernetes 1.24: <release name>"
date: 2022-05-03
slug: kubernetes-1-24-release-announcement
---

# Kubernetes 1.24

**Authors**: [Kubernetes 1.24 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.24/release-team.md): Mickey Boxell (Oracle), Kat Cosgrove (Pulumi), Debabrata Panigrahi, Parthvi Vala (Red Hat)

We are excited to announce the release of Kubernetes 1.24, the first release of 2022!

This release consists of 46 enhancements: 15 enhancements have graduated to stable, 15 enhancements are moving to beta, and 13 enhancements are entering alpha. Also, 2 features have been deprecated, and the [dockershim has been removed](https://kubernetes.io/blog/2022/02/17/dockershim-faq/).     

## Major Themes

### Dockershim Removed from kubelet

> After an initial deprecation in v1.20, the dockershim has been removed from the kubelet. From v1.24 and up, if you are currently relying on Docker Engine as your container runtime, you will need to either use one of the other supported runtimes (such as containerd or CRI-O) or use cri-dockerd. For more information about ensuring your cluster is ready for this removal, please see [this guide](/blog/2022/03/31/ready-for-dockershim-removal/). 

Reviewer: How about:
```suggestion
Following a deprecation period that started with the v1.20 release, Kubernetes no longer
includes the _dockershim_ adapter. If you have been relying on Docker Engine as your
container runtime, you will need to either switch to using a different container runtime
(such as containerd or CRI-O), or use continue to use Docker Engine but via an adapter
such as cri-dockerd.

For information relating the this removal and how it's relevant to upgrades, read
[Is Your Cluster Ready for v1.24?](/blog/2022/03/31/ready-for-dockershim-removal/).
```

(the “1.24 and up” part is implied enough, I hope)

Reviewer: (maybe also reword the release note to match)?

> ### gRPC Probes Graduates to Beta

Reviewer: ```suggestion
### gRPC Probes Graduate to Beta
```
?

With Kubernetes 1.24, the gRPC probes functionality has entered beta and is available by default.

> You can now configure startup, liveness, and readiness probes for your gRPC app without exposing an HTTP endpoint or using an extra executable, natively within Kubernetes.

Reviewer: ```suggestion
You can now [configure startup, liveness, and readiness probes](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes)
for your gRPC app natively within Kubernetes, without exposing an HTTP endpoint or
using an extra executable.
```
(maybe also reword the release note to match)?

### Kubelet Credential Provider graduates to Beta

> Released as Alpha in Kubernetes v1.20, the Kubelet Credential Provider has now graduated to Beta. This allows the kubelet to dynamically retrieve credentials for a container image registry using exec plugins, communicating through stdio using Kubernetes versioned APIs, rather than storing them statically on disk.

Reviewer: We could link to https://kubernetes.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/ here.

Author: Would it be better to link to that or to the issue? https://github.com/kubernetes/enhancements/issues/2133

### Contextual Logging in Alpha
Kubernetes v1.24 has introduced Contextual logging that enables the caller of a function to control all aspects of logging (output formatting, verbosity, additional values and names).

### Avoiding collisions in IP allocation to Services
A Service `ClusterIP` can be assigned:

* dynamically, the cluster will automatically pick a free one within the configured Service IP range.
* statically, the user will set one IP within the configured Service IP range.

Service `ClusterIP` are unique, hence, trying to create a Service with a `ClusterIP` that has already been allocated will return an error.

> Kubernetes 1.24 introduces a new Feature Gate `ServiceIPStaticSubrange` that allows to use a different IP allocation strategy for Services, reducing the risk of collision.

Reviewer: ```suggestion
Kubernetes 1.24 introduces a new way to soft-reserve a range for static IP address assignments
to Services. If you use the new feature (which you must manually enable), your cluster prefers
automatic assignment from the remaining pool of Service IP addresses, reducing the risk of a
collision.
```

## Other Updates
### Graduated to Stable
* To do: Add all enhancements that graduated to stable

### Major Changes
* To do: Add information about Dockershim

### Release Notes
Check out the full details of the Kubernetes 1.24 release in our [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md).
    
### Availability

> Kubernetes 1.24 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.24.0). To get started with Kubernetes, check out these [interactive tutorials](https://kubernetes.io/docs/tutorials/) or run local Kubernetes clusters using Docker container “nodes” with [kind](https://kind.sigs.k8s.io/). You can also easily install 1.24 using [kubeadm](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/). 

Reviewer: `kind` supports more than one container tool, and also I think it might help not to mention Docker here:

```suggestion
Kubernetes 1.24 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.24.0).
To get started with Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run local 
Kubernetes clusters using containers as “nodes”, with [kind](https://kind.sigs.k8s.io/).
You can also easily install 1.24 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/).
```

Any reason we don't also mention minikube or just link to https://kubernetes.io/docs/tasks/tools/ and let readers pick?

* 

### Ecosystem Updates

* KubeCon + CloudNativeCon Europe 2022 will take place in Valencia, Spain, from 16 – 20 May 2022! You can find more information about the conference and registration on the [event site](https://events.linuxfoundation.org/archive/2021/kubecon-cloudnativecon-europe/).
* In the [2021 Cloud Native Survey](https://www.cncf.io/announcements/2022/02/10/cncf-sees-record-kubernetes-and-container-adoption-in-2021-cloud-native-survey/), the CNCF saw record Kubernetes and container adoption. Take a look at the [results of the survey](https://www.cncf.io/reports/cncf-annual-survey-2021/). 

> * The [Linux Foundation](https://www.linuxfoundation.org/) and [The Cloud Native Computing Foundation](https://www.cncf.io/) (CNCF) announced the availability of a new [Cloud Native Developer Bootcamp](https://training.linuxfoundation.org/training/cloudnativedev-bootcamp/?utm_source=lftraining&utm_medium=pr&utm_campaign=clouddevbc0322) to provide participants with the knowledge and skills to design, build, and deploy cloud native applications. Check out [the announcement](https://www.cncf.io/announcements/2022/03/15/new-cloud-native-developer-bootcamp-provides-a-clear-path-to-cloud-native-careers/) to learn more. 

Reviewer: ```suggestion
* The [Linux Foundation](https://www.linuxfoundation.org/) and [The Cloud Native Computing Foundation](https://www.cncf.io/) (CNCF) announced the availability of a new [Cloud Native Developer Bootcamp](https://training.linuxfoundation.org/training/cloudnativedev-bootcamp/?utm_source=lftraining&utm_medium=pr&utm_campaign=clouddevbc0322) to provide participants with the knowledge and skills to design, build, and deploy cloud native applications. Check out the [announcement](https://www.cncf.io/announcements/2022/03/15/new-cloud-native-developer-bootcamp-provides-a-clear-path-to-cloud-native-careers/) to learn more.
```

### Project Velocity

The [CNCF K8s DevStats](https://k8s.devstats.cncf.io/d/12/dashboards?orgId=1&refresh=15m) project aggregates a number of interesting data points related to the velocity of Kubernetes and various sub-projects. This includes everything from individual contributions to the number of companies that are contributing, and is an illustration of the depth and breadth of effort that goes into evolving this ecosystem.

In the v1.24 release cycle, which [ran for 18 weeks](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.24) (January 10 to May 17), we saw contributions from [1029 companies](https://k8s.devstats.cncf.io/d/9/companies-table?orgId=1&var-period_name=v1.23.0%20-%20now&var-metric=contributions) and [1179 individuals](https://k8s.devstats.cncf.io/d/66/developer-activity-counts-by-companies?orgId=1&var-period_name=v1.23.0%20-%20now&var-metric=contributions&var-repogroup_name=Kubernetes&var-country_name=All&var-companies=All&var-repo_name=kubernetes%2Fkubernetes).

## Upcoming Release Webinar

> Join members of the Kubernetes 1.24 release team on <date> to learn about the major features of this release, as well as deprecations and removals to help plan for upgrades. For more information and registration, visit the [event page](#) on the CNCF Online Programs site.

Reviewer: Just calling out that these are placeholders.

## Get Involved
The simplest way to get involved with Kubernetes is by joining one of the many [Special Interest Groups](https://github.com/kubernetes/community/blob/master/sig-list.md) (SIGs) that align with your interests. Have something you’d like to broadcast to the Kubernetes community? Share your voice at our weekly [community meeting](https://github.com/kubernetes/community/tree/master/communication), and through the channels below:

* Find out more about contributing to Kubernetes at the [Kubernetes Contributors](https://www.kubernetes.dev/) website
* Follow us on Twitter [@Kubernetesio](https://twitter.com/kubernetesio) for the latest updates
* Join the community discussion on [Discuss](https://discuss.kubernetes.io/)
* Join the community on [Slack](http://slack.k8s.io/)

> * Post questions (or answer questions) on [Stack Overflow](http://stackoverflow.com/questions/tagged/kubernetes)

Reviewer: ```suggestion
* Post questions (or answer questions) on [Server Fault](https://serverfault.com/questions/tagged/kubernetes).
```
(per a request from the Stack Overflow folks; Stack Overflow aims to be focused on more about writing and debugging code, whereas infrastructure and orchestration questions are fair game for Server Fault).
